### PR TITLE
crimson/seastore: refactor gc into concurrent task

### PR DESCRIPTION
**Pull request by athanatos**

### CODE REVIEW

**Comment by athanatos** [comment]
@liu-chunmei This seems to have fixed several crashes exposed by crimson-store-nbd if you'd like to have a look.

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by athanatos** [comment]
Missed those tests, will fix.

File: src/crimson/os/seastore/transaction_manager.cc [hunk 1]
```diff
@@ -274,69 +290,80 @@ TransactionManager::get_extent_if_live_ret TransactionManager::get_extent_if_liv
...
-		  if (pin->has_been_invalidated() || lref->has_been_invalidated()) {
-		    return crimson::ct_error::eagain::make();
-		  } else {
-		    lref->set_pin(std::move(pin));
-		    lba_manager->add_pin(lref->get_pin());
+    if (is_logical_type(type)) {
+      return lba_manager->get_mapping(
+	t,
+	laddr,
+	len).safe_then([=, &t](lba_pin_list_t pins) {
```
Comment (liu-chunmei) [inline src/crimson/os/seastore/transaction_manager.cc:314]: indent?

File: src/crimson/os/seastore/transaction_manager.cc [hunk 2]
```diff
@@ -274,69 +290,80 @@ TransactionManager::get_extent_if_live_ret TransactionManager::get_extent_if_liv
...
+    } else {
+      logger().debug(
+	"TransactionManager::get_extent_if_live: non-logical extent {}",
+	addr);
+      return lba_manager->get_physical_extent_if_live(
+	t,
+	type,
+	addr,
+	laddr,
+	len);
```
Comment (liu-chunmei) [inline src/crimson/os/seastore/transaction_manager.cc:364]: indent?

File: src/crimson/os/seastore/lba_manager/btree/btree_lba_manager.cc [hunk 3]
```diff
@@ -522,7 +550,11 @@ BtreeLBAManager::update_refcount_ret BtreeLBAManager::update_refcount(
       out.refcount += delta;
       return out;
     }).safe_then([](auto result) {
-      return ref_update_result_t{result.refcount, result.paddr};
+      return ref_update_result_t{
```
[the remark this answers is not in the document]
Comment (athanatos) [reply src/crimson/os/seastore/lba_manager/btree/btree_lba_manager.cc:553]: github is displaying the tabs incorrectly, it's fine.

File: src/crimson/os/seastore/lba_manager/btree/lba_btree_node_impl.cc [hunk 4]
```diff
@@ -169,6 +169,11 @@ LBAInternalNode::mutate_internal_address_ret LBAInternalNode::mutate_internal_ad
     }
     auto iter = get_containing_child(laddr);
     if (iter->get_key() != laddr) {
+      logger().debug(
```
[the remark this answers is not in the document]
Comment (athanatos) [reply src/crimson/os/seastore/lba_manager/btree/lba_btree_node_impl.cc:172]: github is displaying the tabs incorrectly, it's fine.

File: src/crimson/os/seastore/segment_cleaner.cc [hunk 5]
```diff
@@ -197,91 +199,92 @@ void SegmentCleaner::close_segment(segment_id_t segment)
...
 
-SegmentCleaner::do_gc_ret SegmentCleaner::do_gc(
-  Transaction &t,
-  size_t bytes)
+SegmentCleaner::gc_cycle_ret SegmentCleaner::GCProcess::run()
+{
+  return seastar::do_until(
+    [this] { return stopping; },
+    [this] {
+      return maybe_wait_should_run(
```
[the remark this answers is not in the document]
Comment (athanatos) [reply src/crimson/os/seastore/segment_cleaner.cc:229]: github is displaying the tabs incorrectly, it's fine.

File: src/crimson/os/seastore/segment_cleaner.cc [hunk 6]
```diff
@@ -290,53 +293,67 @@ SegmentCleaner::do_gc_ret SegmentCleaner::do_gc(
...
-	  }).safe_then([&t, this] {
-	    if (scan_cursor->is_complete()) {
-	      t.mark_segment_to_release(scan_cursor->get_offset().segment);
-	      mark_releasing(scan_cursor->get_offset().segment);
-	      scan_cursor.reset();
-	    }
-	    return ExtentCallbackInterface::release_segment_ertr::now();
-	  });
+	});
       });
```
[the remark this answers is not in the document]
Comment (athanatos) [reply src/crimson/os/seastore/segment_cleaner.cc:352]: github is displaying the tabs incorrectly, it's fine.

File: src/crimson/os/seastore/segment_cleaner.h [hunk 7]
```diff
@@ -521,22 +539,100 @@ class SegmentCleaner : public JournalSegmentProvider {
...
+    GCProcess(SegmentCleaner &cleaner) : cleaner(cleaner) {}
+
+    void start() {
+      ceph_assert(!process_join);
+      process_join = run();
+    }
+
+    gc_cycle_ret stop() {
+      if (!process_join)
+	return seastar::now();
```
[the remark this answers is not in the document]
Comment (athanatos) [reply src/crimson/os/seastore/segment_cleaner.h:592]: github is displaying the tabs incorrectly, it's fine.

File: src/crimson/os/seastore/segment_cleaner.h [hunk 8]
```diff
@@ -521,22 +539,100 @@ class SegmentCleaner : public JournalSegmentProvider {
...
+	},
+	[this] {
+	  return cleaner.do_gc_cycle();
+	});
+    }
+
+    void maybe_wake_on_space_used() {
+      cleaner.log_gc_state("GCProcess::maybe_wake_on_space_used");
+      if (cleaner.gc_should_run()) {
+	wake();
```
[the remark this answers is not in the document]
Comment (athanatos) [reply src/crimson/os/seastore/segment_cleaner.h:616]: github is displaying the tabs incorrectly, it's fine.

File: src/crimson/os/seastore/transaction_manager.cc [hunk 9]
```diff
@@ -158,7 +160,9 @@ TransactionManager::ref_ret TransactionManager::dec_ref(
       logger().debug(
 	"TransactionManager::dec_ref: offset {} refcount 0",
 	offset);
-      return cache->retire_extent_if_cached(t, result.addr).safe_then([] {
+      return cache->retire_extent_if_cached(
+	t, result.addr, result.length
```
[the remark this answers is not in the document]
Comment (athanatos) [reply src/crimson/os/seastore/transaction_manager.cc:164]: github is displaying the tabs incorrectly, it's fine.

File: src/crimson/os/seastore/transaction_manager.cc (hunk 1 above)
Comment (athanatos) [reply src/crimson/os/seastore/transaction_manager.cc:314]: github is displaying the tabs incorrectly, it's fine.

File: src/crimson/os/seastore/transaction_manager.cc (hunk 2 above)
Comment (athanatos) [reply src/crimson/os/seastore/transaction_manager.cc:364]: github is displaying the tabs incorrectly, it's fine.

**Comment by athanatos** [comment]
@liu-chunmei Each of these files has

```
// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
// vim: ts=8 sw=2 smarttab
```

at the top setting offset and tab width as with most files in the ceph ptoject.  github isn't using the same tab width, so it frequently displays tabs in a way that throws off indents.

File: src/crimson/os/seastore/segment_cleaner.h [hunk 10]
```diff
@@ -601,57 +713,111 @@ class SegmentCleaner : public JournalSegmentProvider {
...
+	get_reclaimable_bytes(),
+	get_reclaim_ratio(),
+	get_available_ratio(),
+	should_block_on_gc(),
+	gc_should_reclaim_space(),
+	journal_head,
+	journal_tail_target,
+	get_dirty_tail(),
+	get_dirty_tail_limit(),
+	gc_should_trim_journal()
```
[the remark this answers is not in the document]
Comment (athanatos) [reply src/crimson/os/seastore/segment_cleaner.h:763]: Most already have comments at their declarations, I'll comment the others.

File: src/crimson/os/seastore/segment_cleaner.cc [hunk 11]
```diff
@@ -290,53 +293,67 @@ SegmentCleaner::do_gc_ret SegmentCleaner::do_gc(
...
+			"SegmentCleaner::gc_reclaim_space: addr {} dead, skipping",
+			addr);
+		      return ExtentCallbackInterface::rewrite_extent_ertr::now();
+		    } else {
+		      logger().debug(
+			"SegmentCleaner::gc_reclaim_space: addr {} alive, gc'ing {}",
+			addr,
+			*ext);
+		    }
+		    return ecb->rewrite_extent(
```
[the remark this answers is not in the document]
Comment (athanatos) [reply src/crimson/os/seastore/segment_cleaner.cc]: Ok

**Comment by athanatos** [comment]
Restricting the total size of journal entries to rewrite seems to have fixed the journal submission crash.  I'm debugging something else now, will push a fix when I figure it out.

**Comment by liu-chunmei** [comment]
> Restricting the total size of journal entries to rewrite seems to have fixed the journal submission crash. I'm debugging something else now, will push a fix when I figure it out.

sure, I have tested the current PR, send you the debug log.

**Comment by athanatos** [comment]
@liu-chunmei The hang I was debugging turned out to be because of a bad_alloc in segment_write -- I'd used --memory 64 instead of --memory 64G so it was using the default memory pool size.  It seems to be stable in my environment at the moment, so I'd like to go ahead and merge this series and take further bugs in a new PR.

**Comment by athanatos** [comment]
I also added an explicit log line for that bad alloc and converted assert() to ceph_assert() in btree_range_pin -- that should make further problems easier to track down.

**Comment by liu-chunmei** [comment]
> @liu-chunmei The hang I was debugging turned out to be because of a bad_alloc in segment_write -- I'd used --memory 64 instead of --memory 64G so it was using the default memory pool size. It seems to be stable in my environment at the moment, so I'd like to go ahead and merge this series and take further bugs in a new PR.

sure, let me test again to make clear the current status. We can merge it first and fix later.

**Review by liu-chunmei — COMMENTED** [review]
Not every time segfault happen when set numjobs=1 and iodepth = xx.(sometimes success, and sometimes fail)  But when set numjobs =2, segfault always happen only once (iodepth =1) success.  no error log output when segfault.  we can fix it later since it can run and not block the further develop work.

**Review by liu-chunmei — APPROVED** [review]

**Pull request merged**